src/processTargets/modifiers/scopeHandlers/BaseScopeHandler.ts

```diff
@@ -0,0 +1,167 @@
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import type { Position, Range, TextEditor } from "vscode";
+import type {
+  Direction,
+  ScopeType,
+} from "../../../typings/targetDescriptor.types";
+import type { TargetScope } from "./scope.types";
+import type {
+  ScopeHandler,
+  ScopeIteratorRequirements,
+} from "./scopeHandler.types";
+import { shouldYieldScope } from "./shouldYieldScope";
+
+/**
+ * All scope handlers should derive from this base class
+ */
+export default abstract class BaseScopeHandler implements ScopeHandler {
+  public abstract readonly scopeType: ScopeType;
+  public abstract readonly iterationScopeType: ScopeType;
+
+  /**
+   * Indicates whether scopes are allowed to contain one another.  If `true`, we
+   * can optimise the algorithm by making certain assumptions.
+   */
+  protected abstract readonly isHierarchical: boolean;
```

Comment:
I mean even if this one is false we can optimize by using `canStopEarly`

Reply:
Whoops; meant to say `false` there, today we only do optimisations for non-hierarchical scopes